docs/examples/schema_ad_hoc.py

```diff
@@ -0,0 +1,20 @@
+from typing import Literal, Union
+
+from typing_extensions import Annotated
+
+from pydantic import BaseModel, Field, schema_json
+
+
+class Cat(BaseModel):
+    pet_type: Literal['cat']
+    cat_name: str
+
+
+class Dog(BaseModel):
+    pet_type: Literal['dog']
+    dog_name: str
+
+
+Pet = Annotated[Union[Cat, Dog], Field(discriminator='pet_type')]
```

Comment:
Why does this need an explicit `discriminator`?

Mypy doesn't need this, so can't we do something similar?

I imagine this as just a `Union[Cat, Dog]`. When you pass something to it, it will try to validate it as `Cat`, if `pet_type != 'cat'` it obviously fails and then tries to validate it as `Dog`. I'm not sure why an explicit discriminator is needed...

Reply:
How is the usage you are imagining different from how things currently work without this discriminator feature?

Reply:
The whole point of discrimination is to try to validate only on matching types hence gaining time on validation and having more explicit error.
Feel free to use plain Union if you want to fallback on other union classes

Reply:
> How is the usage you are imagining different from how things currently work without this discriminator feature?

I think the problem is that I can't use `Pet = Union[Cat, Dog]`, as this is not a `BaseModel`. Maybe this has changed in the new release?

For context, I was sent to this PR in relation to this question: https://github.com/samuelcolvin/pydantic/discussions/2821

Reply:
Why the need of `TypedDict` ?

Reply:
Not sure what you mean? My example just demonstrated how I would achieve tagged unions in normal Python code (with Mypy checking), without Pydantic. I don't understand how to use tagged unions with Pydantic models, that was my question.

Reply:
@Dreamsorcerer I've replied to your question on the issue you linked